Tests for cascading and validation on container elements

gsmet opened the pull request:
This is a big one as things were interleaved in the HV tests. Should be reviewed commit by commit.

* https://hibernate.atlassian.net/browse/BVTCK-135
* https://hibernate.atlassian.net/browse/BVTCK-132 (addition to an already fixed issue)
* https://hibernate.atlassian.net/browse/BVTCK-141
* https://hibernate.atlassian.net/browse/BVTCK-140
* https://hibernate.atlassian.net/browse/BVTCK-137
* https://hibernate.atlassian.net/browse/BVTCK-136
* https://hibernate.atlassian.net/browse/BVTCK-152

gunnarmorling commented on tests/src/main/java/org/hibernate/beanvalidation/tck/tests/validation/graphnavigation/containerelement/LegacyValidOnContainerCascadingTest.java:
```diff
@@ -0,0 +1,304 @@
+
+/**
+ * Test that the legacy behavior of having @Valid on the container works correctly: it validates the elements of the
+ * iterable AND the properties of the container if any.
```
I would prefer to _not_ test that latter part in the TCK. The reason being that we don't clearly state that it should happen:

> [@Valid] causes the contents of the iterator to be validated [...] Each object provided by the iterator is validated. For Map, the value (retrieved by getValue) of each Map.Entry is validated (the key is not validated).

So what HV is doing, goes beyond the spec, but I don't think it can be considered portable behavior and thus can be tested in the TCK.

gsmet replied:
Mmmmh, OK. I must admit I think it's the normal behavior. But I can remove this part for sure.

gunnarmorling commented on tests/src/main/java/org/hibernate/beanvalidation/tck/tests/validation/graphnavigation/containerelement/model/VisitorWithGroups.java:
```diff
@@ -0,0 +1,25 @@
+	public VisitorWithGroups(String name) {
+		this.name = name;
+	}
+}
```
An empty line is missing.

gunnarmorling commented on tests/src/main/java/org/hibernate/beanvalidation/tck/tests/valueextraction/resolution/ValueExtractorResolutionAlgorithmTest.java:
```diff
@@ -0,0 +1,49 @@
+				.build();
+	}
+
+	@Test(expectedExceptions = ConstraintDeclarationException.class)
```
Do we somewhere have a test for `ValueExtractorDeclarationException` and `ValueExtractorDefinitionException`?

gsmet replied:
We have some tests for `ValueExtractorDefinitionException`. The `ValueExtractorDeclarationException` are still missing. I opened issues last week with all the missing parts.

gunnarmorling commented on tests/src/main/java/org/hibernate/beanvalidation/tck/tests/constraints/containerelement/ContainerElementConstraintListTest.java:
```diff
@@ -0,0 +1,307 @@
+	private static class TypeWithList5 {
+
+		@SuppressWarnings("unused")
+		public void setValues(@Valid List<@NotNull @NotBlank String> listParameter) {
```
Is it intentional that you use the legacy style here?

gsmet replied:
Argh, no. The HV tests were doing it wrong. I fixed it everywhere but missed this one. Will fix.

gunnarmorling commented on tests/src/main/java/org/hibernate/beanvalidation/tck/tests/constraints/containerelement/ContainerElementConstraintMapKeyTest.java:
```diff
@@ -0,0 +1,309 @@
+	private static class TypeWithMap5 {
+
+		@SuppressWarnings("unused")
+		public void setValues(@Valid Map<@NotNull @NotBlank String, String> mapParameter) {
```
Same question about using the legacy style. Is it even needed here at all given that this test is about the map key?

gunnarmorling commented on tests/src/main/java/org/hibernate/beanvalidation/tck/tests/constraints/containerelement/ContainerElementConstraintMapValueTest.java:
```diff
@@ -0,0 +1,309 @@
+	private static class TypeWithMap5 {
+
+		@SuppressWarnings("unused")
+		public void setValues(@Valid Map<String, @NotNull @NotBlank String> mapParameter) {
```
Same question about legacy style as above.

gunnarmorling commented on tests/src/main/java/org/hibernate/beanvalidation/tck/tests/constraints/containerelement/ContainerElementConstraintOptionalTest.java:
```diff
@@ -0,0 +1,221 @@
+		private Optional<String> stringOptional;
+
+		@SuppressWarnings("unused")
+		public Optional<@NotNull @NotBlank String> returnStringOptional() {
```
An optional of not null is a bit painful. Use another constraint perhaps? Do we have one test on the payload business btw. (applying it to the wrapper)?

gsmet replied:
No, we don't have them yet. There's an issue for it.

gunnarmorling commented on tests/src/main/java/org/hibernate/beanvalidation/tck/tests/constraints/containerelement/ContainerElementConstraintSetTest.java:
```diff
@@ -0,0 +1,310 @@
+
+	private static class TypeWithSet2 {
+
+		@Valid
```
Seems a global check on using legacy style in this PR is in order (unless it's intentional, but I reckon it stems from copying existing tests).

gsmet replied:
Yeah, I fixed it in some of the tests but apparently missed some. Will make a global pass on all `@Valid` definitions.

gunnarmorling commented on tests/src/main/java/org/hibernate/beanvalidation/tck/tests/constraints/containerelement/ContainerElementConstraintSetTest.java:
```diff
@@ -0,0 +1,310 @@
+		public TypeWithSet6(@Valid Set<@NotNull @NotBlank String> setParameter) {
+		}
+	}
+}
```
No empty line at EOF.

gunnarmorling commented on tests/src/main/java/org/hibernate/beanvalidation/tck/tests/constraints/containerelement/NestedContainerElementConstraintsTest.java:
```diff
@@ -0,0 +1,297 @@
+						.propertyPath( pathWith()
+								.property( "map" )
+								.containerElement( "<map value>", true, "key1", null, Map.class, 1 )
+								.containerElement( "<list element>", true, null, 0, List.class, 0 )
```
Was wondering why there's no element for the `Optional`, but we don't add a node I'm realizing now.

gsmet replied:
Yup!

gsmet commented:
@gunnarmorling addressed your comments.

About the legacy tests, not sure if it wouldn't be best to improve the spec. Considering it's a legacy behavior, I can see how it's not appealing.

Anyway, I removed this part from this PR.

gunnarmorling commented:
Rebased and applied; that was a big one, thanks a lot, @gsmet!
